src/System.Private.CoreLib/src/System/StartupHookProvider.cs

```diff
@@ -0,0 +1,121 @@
+        private static void CallStartupHook(string assemblyPath)
+        {
+            Debug.Assert(!String.IsNullOrEmpty(assemblyPath));
+            Debug.Assert(!PathInternal.IsPartiallyQualified(assemblyPath));
+
+            Assembly assembly = AssemblyLoadContext.Default.LoadFromAssemblyPath(assemblyPath);
+            Debug.Assert(assembly != null);
+            Type type = assembly.GetType(StartupHookTypeName, throwOnError: true);
+
+            // Look for a static method without any parameters
+            MethodInfo initializeMethod = type.GetMethod(InitializeMethodName,
+                                                         BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static,
+                                                         null, // use default binder
+                                                         Type.EmptyTypes, // parameters
+                                                         null); // no parameter modifiers
+
+            bool wrongSignature = false;
+            if (initializeMethod == null)
+            {
+                // There weren't any static methods without
+                // parameters. Look for any methods with the correct
+                // name, to provide precise error handling.
+                try
+                {
+                    // This could find zero, one, or multiple methods
+                    // with the correct name.
+                    initializeMethod = type.GetMethod(InitializeMethodName,
+                                                      BindingFlags.Public | BindingFlags.NonPublic |
+                                                      BindingFlags.Static | BindingFlags.Instance);
+                }
+                catch (AmbiguousMatchException)
+                {
+                    // Found multiple
+                    Debug.Assert(initializeMethod == null);
+                    wrongSignature = true;
+                }
+                if (initializeMethod != null)
+                {
+                    // Found one
+                    wrongSignature = true;
+                }
+                else
+                {
+                    // Didn't find any
+                    throw new MissingMethodException(StartupHookTypeName, InitializeMethodName);
+                }
+            }
+            else if (initializeMethod.ReturnType != typeof(void))
+            {
+                wrongSignature = true;
+            }
+
+            if (wrongSignature)
+            {
+                throw new ArgumentException(SR.Format(SR.Argument_InvalidStartupHookSignature,
+                                                      StartupHookTypeName + Type.Delimiter + InitializeMethodName,
+                                                      assemblyPath));
+            }
+
+            Debug.Assert(initializeMethod != null &&
+                         (initializeMethod.IsPublic || initializeMethod.IsPrivate) &&
```

Comment:
This is not covering all non-public methods. Non-public methods include protected, etc. I think it is best to delete this part of the assert.